### PR TITLE
feat(psl): Recommend type in error message if schema validation fails because of case

### PR DESCRIPTION
**Pull request by pranayat**
Hi!
This PR closes https://github.com/prisma/prisma/issues/15174
For now I've fixed this issue for 'ScalarType' types and titled the PR 'WIP' so you can let me know if I'm doing something absolutely inane. I've added a test as well.
Thanks!

### CODE REVIEW

File: psl/diagnostics/src/error.rs [hunk 1]
```diff
@@ -293,6 +293,13 @@ impl DatamodelError {
         Self::new(msg, span)
     }
 
+    pub fn new_type_for_case_not_found_error(type_name: &str, suggestion: &str, span: Span) -> DatamodelError {
+        let msg = format!(
+            "Type \"{type_name}\" is neither a built-in type, nor refers to another model, custom type, or enum. Did you mean {suggestion}?"
```
Comment (pranayat) [inline psl/diagnostics/src/error.rs:298]: Although, I've only checked for built-in types for now.

File: psl/parser-database/src/types.rs [hunk 2]
```diff
@@ -1423,18 +1434,33 @@ impl ScalarType {
...
-            "String" => Some(ScalarType::String),
-            "DateTime" => Some(ScalarType::DateTime),
-            "Json" => Some(ScalarType::Json),
-            "Bytes" => Some(ScalarType::Bytes),
-            "Decimal" => Some(ScalarType::Decimal),
-            _ => None,
+    pub(crate) fn try_from_str(s: &str, ignore_case: bool) -> Option<ScalarType> {
+        if ignore_case {
+            match s.to_lowercase().as_str() {
+                "int"  => Some(ScalarType::Int),
```
Comment (pranayat) [inline psl/parser-database/src/types.rs:1440]: remove space

File: psl/parser-database/src/types.rs [hunk 3]
```diff
@@ -1423,18 +1434,33 @@ impl ScalarType {
...
+                "bigint" => Some(ScalarType::BigInt),
+                "float" => Some(ScalarType::Float),
+                "boolean" => Some(ScalarType::Boolean),
+                "string" => Some(ScalarType::String),
+                "datetime" => Some(ScalarType::DateTime),
+                "json" => Some(ScalarType::Json),
+                "bytes" => Some(ScalarType::Bytes),
+                "decimal" => Some(ScalarType::Decimal),
+                _ => None,
+            }
```
Comment (pranayat) [inline psl/parser-database/src/types.rs:1450]: remove space

**Comment by pranayat** [comment]
(Force)pushed an update that tries to match a model's field type against scalar types as well as names of all top level entries in the schema definition.

**Comment by pranayat** [comment]
I should probably also exclude `Source` and `Generator` top level entries when looking for suggestions, i.e. only look for suggestions in `Enum`, `CompositeType` and `Model` entry names, since only these types can be referenced as model fields.

**Comment by pranayat** [comment]
No cigar ? 😅

**Comment by Druue** [comment]
Hey, @pranayat 

Thanks for submitting!

https://github.com/prisma/prisma-engines/assets/29753584/01f6ce0b-9c56-435f-8dc9-70527a80f368

---

> Could it make sense to add a function similar to iter_tops() in ast.rs that returns only these top level entries ?

Yeah, absolutely it would!
We've been using the following naming in Language-Tools:
```ts
export type ConfigBlockType = 'generator' | 'datasource'

export type DatamodelBlockType = 'model' | 'view' | 'type' | 'enum'

export type BlockType = ConfigBlockType | DatamodelBlockType
```

So perhaps something along the lines of `iter_datamodels`

**Comment by pranayat** [comment]
@Druue Sounds good 😄
I've made the change and updated the test.

**Review by Druue — APPROVED** [review]
✨

File: psl/diagnostics/src/error.rs (hunk 1 above)
Comment (janpio) [inline psl/diagnostics/src/error.rs:298]:
```suggestion
            "Type \"{type_name}\" is neither a built-in type, nor refers to another model, custom type, or enum. Did you mean \"{suggestion}\"?"
```

File: psl/psl/tests/base/basic.rs [hunk 4]
```diff
@@ -239,3 +239,67 @@ fn type_aliases_must_error() {
...
+        [1;94m 4 | [0mmodel B {
+        [1;94m 5 | [0m  a  [1;91mdatetime[0m
+        [1;94m   | [0m
+        [1;91merror[0m: [1mType "footime" is neither a built-in type, nor refers to another model, custom type, or enum.[0m
+          [1;94m-->[0m  [4mschema.prisma:6[0m
+        [1;94m   | [0m
+        [1;94m 5 | [0m  a  datetime
+        [1;94m 6 | [0m  b  [1;91mfootime[0m
+        [1;94m   | [0m
+        [1;91merror[0m: [1mType "user" is neither a built-in type, nor refers to another model, custom type, or enum. Did you mean User?[0m
```
Comment (janpio) [inline psl/psl/tests/base/basic.rs:284]:
```suggestion
        [1;91merror[0m: [1mType "datetime" is neither a built-in type, nor refers to another model, custom type, or enum. Did you mean "DateTime"?[0m
          [1;94m-->[0m  [4mschema.prisma:5[0m
        [1;94m   | [0m
        [1;94m 4 | [0mmodel B {
        [1;94m 5 | [0m  a  [1;91mdatetime[0m
        [1;94m   | [0m
        [1;91merror[0m: [1mType "footime" is neither a built-in type, nor refers to another model, custom type, or enum.[0m
          [1;94m-->[0m  [4mschema.prisma:6[0m
        [1;94m   | [0m
        [1;94m 5 | [0m  a  datetime
        [1;94m 6 | [0m  b  [1;91mfootime[0m
        [1;94m   | [0m
        [1;91merror[0m: [1mType "user" is neither a built-in type, nor refers to another model, custom type, or enum. Did you mean "User"?[0m
```

Comment (pranayat) [reply psl/psl/tests/base/basic.rs]: Updated :)

**Pull request merged**